fix: Route Builder rules should be case insensitive

jemiluv8 opened the pull request:
## What does this PR do?
Override jsonLogic operators on string operands to allow for case insensitive comparisons

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes #5439

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->
[Case insensitive "equals" comparison](https://www.loom.com/share/c03892743934450aba0e76d5a5a97a4d)
[Case insensitive "not equals" comparison](https://www.loom.com/share/9a15da4043324d6aa974fdaa8267d975)

[Case insensitive "contains" comparison](https://www.loom.com/share/61e636b6e11b4350831e67027fea5e3c)
[Case insensitive "not contains" comparison](https://www.loom.com/share/c8f4402e6cc74875a0d6ecd07bd922d7)

## Type of change

<!-- Please delete bullets that are not relevant. -->

- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [ ] Install the "Routing Forms" app
- [ ] Create a new form
- [ ] Create form fields under the "Form" tab
- [ ] Go to the routing tab and create a rule for any of the string operators ("contains", "not contains", "equals", "not equals")
- [ ] Use the test preview on the left sidebar to fill out an example form
- [ ] Use the "Test Routing" button to confirm the case insensitive comparison of the form values for relevant rules

Edit by @hariombalhara 
- We haven't implemented the same for Reporting because [Prisma as of this moment doesn't support case insensitive JSON querying](https://github.com/prisma/prisma/issues/7390).

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->
- I haven't added tests that prove my fix is effective or that my feature works

/claim #5439

jemiluv8 committed:
override jsonLogic operators on string operands to allow for case insensitive comparisons.

Affected Operators: "==", "===", "!=", "!==", "in"

jemiluv8 committed:
Merge remote-tracking branch 'origin' into CAL-336-Route-builder-rules-should-be-case-insensitive

jemiluv8 committed:
disable no-explicit-any on jsonLogicOverrides file since most of the code there will be from jsonLogic and may not meet our coding style.

Majority of overrides require us to copy over functions and their signatures from jsonLogic and then modify their implementation.

The signature of functions implementing most operators take the operands typed as "any", which is intended, but doesn't adhere to our coding style. Hence the need to override the eslint rule

jemiluv8 committed:
run linter to fix issues

hariombalhara committed:
Merge branch 'main' into CAL-336-Route-builder-rules-should-be-case-insensitive

hariombalhara commented on packages/app-store/routing-forms/lib/jsonLogicOverrides.ts:
```diff
@@ -0,0 +1,32 @@
+jsonLogic.add_operation("in", function (a: any, b: any) {
+  const first = normalize(a);
+  const second = normalize(b);
+  if (!second || typeof second.indexOf === "undefined") return false;
```
What would be the case when second.indexOf is undefined. As per my testing second is either an array of string or string. I couldn't find such a case.

jemiluv8 replied:
The "in" function was copied over from the jsonLogic github repository. 
My intention was to not change anything besides normalizing the input.

I assumed it was to cater for a corner case but will review and decide whether
to modify it with the expectation of second being string or string[] as you found
out from your tests.

hariombalhara committed:
Fix bug in in operator when second arg is an array

hariombalhara commented on packages/app-store/routing-forms/lib/jsonLogicOverrides.ts:
```diff
@@ -0,0 +1,53 @@
+  if (typeof input === "string") {
+    return input.toLowerCase() as T;
+  }
+  if (input instanceof Array) {
```
A failing test caught this bug where input can sometimes be an array as well(for MultiSelect it is)

jemiluv8 replied:
Got it.

hariombalhara reviewed (changes requested):
Looks good so far except 1 bug where MultiSelect equals stopped working.

That was because 'in' operator got second argument as an array of string which wasn't handled in normalize. So, one operand got lowercased but not the other one causing in operator to always fail. I have fixed that case.

@jemiluv8 I would suggest you to do one more round of testing with all the operators and field types combination.

Also, there is an open comment.

jemiluv8 committed:
remove redundant indexOf check on overriden jsonLogic "in" operator.

Note: this deviates from the original implementation in the jsonLogic library because our current useage ensures that the second operand is always a string or string[] and will therefore always have .index function. Whenever our invariants change in the future, make sure to modify this implementation to prevent any unexpected

jemiluv8 commented:
@hariombalhara, I just completed another round of testing and all looks fine.
I admit I didn't pay much attention to select/multiselect field types and their
respective operations

**About the redundant check in our override for the "in" operator**
I decided to remove the redundant check for the ```typeof second.indexOf === "undefined"```.

Like I mentioned in the comment above I tried to copy over the exact implementation of
various jsonLogic operators from the repo. For the "in" operator, the library assumed the 
second operand could be of any type and was therefore guarding against situations where
it may not be a string or an array.

Our current useage of the jsonLogic library ensures that the "in" operator will always get either a
string or string[]. That makes the extra check redundant.

hariombalhara commented on playwright.config.ts:
```diff
@@ -11,12 +11,12 @@ const outputDir = path.join(__dirname, "test-results");
-const DEFAULT_NAVIGATION_TIMEOUT = process.env.CI ? 15000 : 50000;
-const DEFAULT_EXPECT_TIMEOUT = process.env.CI ? 15000 : 50000;
+const DEFAULT_NAVIGATION_TIMEOUT = process.env.CI ? 15000 : 120000;
+const DEFAULT_EXPECT_TIMEOUT = process.env.CI ? 15000 : 120000;
```
Added this change here because sometimes dev server can be too slow and it times out unncessarily.
